File: src/services/activity.ts
import {
  collection,
  addDoc,
  getDocs,
  query,
  where,
  orderBy,
  limit,
  Timestamp
} from 'firebase/firestore';
import { db, auth } from '@/lib/firebase';
import { getFollowing } from './follows';

export interface Activity {
  id?: string;
  userId: string;
  username: string;
  userAvatar?: string;
  type: 'log' | 'review' | 'like' | 'list' | 'follow';
  targetId: string;
  targetType: 'raceLog' | 'list' | 'user';
  content?: string;
  createdAt: Date;
}

const activitiesCollection = collection(db, 'activities');

export const createActivity = async (activity: Omit<Activity, 'id' | 'createdAt' | 'userId' | 'username' | 'userAvatar'>) => {
  const user = auth.currentUser;
  if (!user) throw new Error('User not authenticated');

  const newActivity = {
    ...activity,
    userId: user.uid,
    username: user.displayName || user.email?.split('@')[0] || 'User',
    userAvatar: user.photoURL || '',
    createdAt: Timestamp.now()
  };

  const docRef = await addDoc(activitiesCollection, newActivity);
  return docRef.id;
};

export const getUserActivity = async (userId: string, limitCount: number = 20) => {
  const q = query(
    activitiesCollection,
    where('userId', '==', userId),
    orderBy('createdAt', 'desc'),
    limit(limitCount)
  );
  const snapshot = await getDocs(q);
  return snapshot.docs.map(doc => ({ id: doc.id, ...doc.data() } as Activity));
};

export const getFollowingActivity = async (limitCount: number = 50) => {
  const user = auth.currentUser;
  if (!user) throw new Error('User not authenticated');

  const following = await getFollowing(user.uid);
  const followingIds = following.map(f => f.followingId);

  if (followingIds.length === 0) {
    return [];
  }

  const q = query(
    activitiesCollection,
    where('userId', 'in', followingIds.slice(0, 10)),
    orderBy('createdAt', 'desc'),
    limit(limitCount)
  );
  const snapshot = await getDocs(q);
  return snapshot.docs.map(doc => ({ id: doc.id, ...doc.data() } as Activity));
};

export const getGlobalActivity = async (limitCount: number = 50) => {
  const q = query(
    activitiesCollection,
    orderBy('createdAt', 'desc'),
    limit(limitCount)
  );
  const snapshot = await getDocs(q);
  return snapshot.docs.map(doc => ({ id: doc.id, ...doc.data() } as Activity));
};

File: lib/core/models/activity.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class Activity {
  final String? id;
  final String userId;
  final String username;
  final String? userAvatar;
  final String type; // 'log', 'review', 'like', 'list', 'follow'
  final String targetId;
  final String targetType; // 'raceLog', 'list', 'user'
  final String? content;
  final DateTime createdAt;

  Activity({
    this.id,
    required this.userId,
    required this.username,
    this.userAvatar,
    required this.type,
    required this.targetId,
    required this.targetType,
    this.content,
    required this.createdAt,
  });

  factory Activity.fromJson(Map<String, dynamic> json, String id) {
    return Activity(
      id: id,
      userId: json['userId'] ?? '',
      username: json['username'] ?? 'User',
      userAvatar: json['userAvatar'],
      type: json['type'] ?? 'log',
      targetId: json['targetId'] ?? '',
      targetType: json['targetType'] ?? 'raceLog',
      content: json['content'],
      createdAt: (json['createdAt'] as Timestamp).toDate(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'userId': userId,
      'username': username,
      'userAvatar': userAvatar,
      'type': type,
      'targetId': targetId,
      'targetType': targetType,
      'content': content,
      'createdAt': Timestamp.fromDate(createdAt),
    };
  }
}
